crates/burn-core/src/module/param/base.rs

```diff
@@ -1,40 +1,265 @@
 use super::ParamId;
+use alloc::boxed::Box;
 use alloc::format;
+use burn_common::stub::{RwLock, SyncOnceCell};
+use core::ops::Deref;
 
-/// Define a parameter.
-#[derive(new, Debug, Clone)]
-pub struct Param<T> {
+/// Parameters are the fundamental building blocks of [modules](crate::module::Module) where they
+/// serve as containers for [tensors](crate::tensor::Tensor) that can be updated during
+/// training, and loaded during inference. If you don't want to save the tensors with a record
+/// and/or don't want to update it during training, you don't need this type to wrap your tensor.
+///
+/// # Laziness
+///
+/// The initialization of parameters can be lazy when created using
+/// [uninitialized](Self::uninitialized), which can be done using an [initializer](crate::nn::Initializer).
+///
+/// This reduces the amount of allocations done when loading a model for inference without having
+/// to create a custom initialization function only for inference.
+///
+/// ## Example
+///
+/// ```rust, ignore
+/// let device = Device::default();
+/// let config = ModuleConfig::default();
+/// let record = Recorder::new().load("/path/to/module", &device);
+///
+/// // No tensor allocation
+/// let module = config.init(device);
+/// // Will use the tensor allocated for the record if the same device is used.
+/// let module = module.load_record(record);
+/// ```
+pub struct Param<T: Parameter> {
     pub(crate) id: ParamId,
-    pub(crate) value: T,
+    state: SyncOnceCell<T>,
+    /// The locking is only required because of `lazy_device` and `lazy_is_require_grad`.
+    ///
+    /// Because of [OnceCell](OnceCell), we have a guarantee that the initialization will only be called once,
+    /// but it may be called at the same time as `lazy_device` and `lazy_is_require_grad`, which is
+    /// when the lock is actually useful, waiting for the initialization to be completed before
+    /// returning the value.
+    ///
+    /// To avoid creating locks on already initialized parameter, we wrap the lock inside an
+    /// Option, the inner option is required for resetting the state onced initialized.
+    /// TLDR: RwLock(None) only happens on the param reference that is lazy, but was initialized,
+    /// all other parameters
```

Comment:
I can't parse this sentence

Reply:
I feel like the last sentence is incomplete 😅

all other parameters [blank].. what?